From 10a7295a871f3894165c561f0d669f03702d7f3f Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 4 Aug 2026 16:09:47 -0400
Subject: [PATCH 258/268] Revert "ntdll: Support x86_64 syscall emulation."

This reverts commit abd586c858c23190d847abb71fe04d19af718c70.
---
 configure.ac                    |   1 -
 dlls/ntdll/unix/signal_x86_64.c | 195 +-------------------------------
 2 files changed, 1 insertion(+), 195 deletions(-)

diff --git a/configure.ac b/configure.ac
index 1da63b121a2..e54b6cd4d52 100644
--- a/configure.ac
+++ b/configure.ac
@@ -692,7 +692,6 @@ AC_CHECK_HEADERS(\
 	linux/memfd.h \
 	linux/ntsync.h \
 	linux/param.h \
-	linux/seccomp.h \
 	linux/serial.h \
 	linux/types.h \
 	linux/ucdrom.h \
diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index 110404c274c..4c226212a9f 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -27,7 +27,6 @@
 #include "config.h"
 
 #include <assert.h>
-#include <errno.h>
 #include <pthread.h>
 #include <signal.h>
 #include <stdlib.h>
@@ -35,8 +34,6 @@
 #include <stdio.h>
 #include <sys/types.h>
 #include <sys/mman.h>
-#include <sys/stat.h>
-#include <fcntl.h>
 #include <unistd.h>
 #ifdef HAVE_MACHINE_SYSARCH_H
 # include <machine/sysarch.h>
@@ -76,14 +73,6 @@
 extern void _thread_set_tsd_base(uint64_t);
 #endif
 
-#if defined(HAVE_LINUX_FILTER_H) && defined(HAVE_LINUX_SECCOMP_H) && defined(HAVE_SYS_PRCTL_H)
-#define HAVE_SECCOMP 1
-# include <linux/filter.h>
-# include <linux/seccomp.h>
-# include <sys/prctl.h>
-# include <linux/audit.h>
-#endif
-
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "windef.h"
@@ -924,8 +913,7 @@ static inline ucontext_t *init_handler( void *sigcontext )
     if (fs32_sel)
     {
         struct ntdll_thread_data *thread_data = (struct ntdll_thread_data *)&get_current_teb()->GdiTebBatch;
-        if (((struct amd64_thread_data *)thread_data->cpu_data)->pthread_teb)
-            arch_prctl( ARCH_SET_FS, ((struct amd64_thread_data *)thread_data->cpu_data)->pthread_teb );
+        arch_prctl( ARCH_SET_FS, ((struct amd64_thread_data *)thread_data->cpu_data)->pthread_teb );
     }
 #elif defined __APPLE__
     {
@@ -2043,186 +2031,6 @@ static inline DWORD is_privileged_instr( CONTEXT *context )
     return 0;
 }
 
-#ifdef HAVE_SECCOMP
-static void sigsys_handler( int signal, siginfo_t *siginfo, void *sigcontext )
-{
-    extern const void *__wine_syscall_dispatcher_prolog_end_ptr;
-    ucontext_t *ucontext = init_handler( sigcontext );
-    struct syscall_frame *frame = get_syscall_frame();
-
-    TRACE_(seh)("SIGSYS, rax %#llx, rip %#llx.\n", RAX_sig(ucontext), RIP_sig(ucontext));
-
-    if (RAX_sig(ucontext) == 0xffff)
-    {
-        /* Test syscall from the Unix side (install_bpf). */
-        RAX_sig(ucontext) = STATUS_INVALID_PARAMETER;
-        return;
-    }
-
-    frame->rip = RIP_sig(ucontext) + 0xb;
-    frame->rcx = RIP_sig(ucontext);
-    frame->eflags = EFL_sig(ucontext);
-    frame->restore_flags = 0;
-    if (instrumentation_callback) frame->restore_flags |= RESTORE_FLAGS_INSTRUMENTATION;
-    RCX_sig(ucontext) = (ULONG_PTR)frame;
-    R11_sig(ucontext) = frame->eflags;
-    if (EFL_sig(ucontext) & 0x100)
-    {
-        EFL_sig(ucontext) &= ~0x100;  /* clear single-step flag */
-        frame->restore_flags |= CONTEXT_CONTROL;
-    }
-    RIP_sig(ucontext) = (ULONG64)__wine_syscall_dispatcher_prolog_end_ptr;
-}
-#endif
-
-#ifdef HAVE_SECCOMP
-static int sc_seccomp(unsigned int operation, unsigned int flags, void *args)
-{
-#ifndef __NR_seccomp
-#   define __NR_seccomp 317
-#endif
-    return syscall(__NR_seccomp, operation, flags, args);
-}
-#endif
-
-static void check_bpf_jit_enable(void)
-{
-    char enabled;
-    int fd;
-
-    fd = open("/proc/sys/net/core/bpf_jit_enable", O_RDONLY);
-    if (fd == -1)
-    {
-        WARN_(seh)("Could not open /proc/sys/net/core/bpf_jit_enable.\n");
-        return;
-    }
-
-    if (read(fd, &enabled, sizeof(enabled)) == sizeof(enabled))
-    {
-        TRACE_(seh)("enabled %#x.\n", enabled);
-
-        if (enabled != '1')
-            ERR_(seh)("BPF JIT is not enabled in the kernel, enable it to reduce syscall emulation overhead.\n");
-    }
-    else
-    {
-        WARN_(seh)("Could not read /proc/sys/net/core/bpf_jit_enable.\n");
-    }
-    close(fd);
-}
-
-static void install_bpf(struct sigaction *sig_act)
-{
-#ifdef HAVE_SECCOMP
-#   ifndef SECCOMP_FILTER_FLAG_SPEC_ALLOW
-#       define SECCOMP_FILTER_FLAG_SPEC_ALLOW (1UL << 2)
-#   endif
-
-#   ifndef SECCOMP_SET_MODE_FILTER
-#       define SECCOMP_SET_MODE_FILTER 1
-#   endif
-    static const BYTE syscall_trap_test[] =
-    {
-        0x48, 0x89, 0xf8,   /* mov %rdi, %rax */
-        0x0f, 0x05,         /* syscall */
-        0xc3,               /* retq */
-    };
-    static const unsigned int flags = SECCOMP_FILTER_FLAG_SPEC_ALLOW;
-
-#define NATIVE_SYSCALL_ADDRESS_START 0x700000000000
-
-    static struct sock_filter filter[] =
-    {
-        BPF_STMT(BPF_LD | BPF_W | BPF_ABS, offsetof(struct seccomp_data, instruction_pointer) + 4),
-        /* Native libs are loaded at high addresses. */
-        BPF_JUMP(BPF_JMP | BPF_JGT | BPF_K, NATIVE_SYSCALL_ADDRESS_START >> 32, 0, 1),
-        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
-        /* Allow i386. */
-        BPF_STMT(BPF_LD | BPF_W | BPF_ABS, offsetof(struct seccomp_data, arch)),
-        BPF_JUMP (BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_X86_64, 1, 0),
-        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
-        /* Allow wine64-preloader */
-        BPF_STMT(BPF_LD | BPF_W | BPF_ABS, offsetof(struct seccomp_data, instruction_pointer)),
-        BPF_JUMP(BPF_JMP | BPF_JGE | BPF_K, 0x7d400000, 1, 0),
-        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_TRAP),
-        BPF_JUMP(BPF_JMP | BPF_JGE | BPF_K, 0x7d402000, 0, 1),
-        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_TRAP),
-        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
-    };
-    long (*test_syscall)(long sc_number);
-    struct sock_fprog prog;
-    NTSTATUS status;
-
-    if ((ULONG_PTR)sc_seccomp < NATIVE_SYSCALL_ADDRESS_START
-            || (ULONG_PTR)syscall < NATIVE_SYSCALL_ADDRESS_START)
-    {
-        ERR_(seh)("Native libs are being loaded in low addresses, sc_seccomp %p, syscall %p, not installing seccomp.\n",
-                sc_seccomp, syscall);
-        ERR_(seh)("The known reasons are /proc/sys/vm/legacy_va_layout set to 1 or 'ulimit -s' being 'unlimited'.\n");
-        return;
-    }
-
-    sig_act->sa_sigaction = sigsys_handler;
-    memset(&prog, 0, sizeof(prog));
-
-    sigaction(SIGSYS, sig_act, NULL);
-
-    test_syscall = mmap((void *)0x600000000000, 0x1000, PROT_EXEC | PROT_READ | PROT_WRITE,
-            MAP_FIXED_NOREPLACE | MAP_PRIVATE | MAP_ANON, -1, 0);
-    if (test_syscall != (void *)0x600000000000)
-    {
-        int ret;
-
-        ERR("Could not allocate test syscall, falling back to seccomp presence check, test_syscall %p, errno %d.\n",
-                test_syscall, errno);
-        if (test_syscall != MAP_FAILED) munmap(test_syscall, 0x1000);
-
-        if ((ret = prctl(PR_GET_SECCOMP, 0, NULL, 0, 0)))
-        {
-            if (ret == 2)
-                TRACE_(seh)("Seccomp filters already installed.\n");
-            else
-                ERR_(seh)("Seccomp filters cannot be installed, ret %d, error %s.\n", ret, strerror(errno));
-            return;
-        }
-    }
-    else
-    {
-        memcpy(test_syscall, syscall_trap_test, sizeof(syscall_trap_test));
-        status = test_syscall(0xffff);
-        munmap(test_syscall, 0x1000);
-        if (status == STATUS_INVALID_PARAMETER)
-        {
-            TRACE_(seh)("Seccomp filters already installed.\n");
-            return;
-        }
-        if (status != -ENOSYS && (status != -1 || errno != ENOSYS))
-        {
-            ERR_(seh)("Unexpected status %#x, errno %d.\n", status, errno);
-            return;
-        }
-    }
-
-    TRACE_(seh)("Installing seccomp filters.\n");
-
-    prog.len = ARRAY_SIZE(filter);
-    prog.filter = filter;
-
-    if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0))
-    {
-        ERR_(seh)("prctl(PR_SET_NO_NEW_PRIVS, ...): %s.\n", strerror(errno));
-        return;
-    }
-    if (sc_seccomp(SECCOMP_SET_MODE_FILTER, flags, &prog))
-    {
-        ERR_(seh)("prctl(PR_SET_SECCOMP, ...): %s.\n", strerror(errno));
-        return;
-    }
-    check_bpf_jit_enable();
-#else
-    WARN_(seh)("Built without seccomp.\n");
-#endif
-}
 
 /***********************************************************************
  *           handle_interrupt
@@ -3187,7 +2995,6 @@ void signal_init_process(void)
     sig_act.sa_sigaction = sigsys_handler;
     if (sigaction( SIGSYS, &sig_act, NULL ) == -1) goto error;
 #endif
-    install_bpf(&sig_act);
     return;
 
  error:
-- 
2.55.0

